### PR TITLE
add libtorch/1.8.1

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **lib/1.0**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

https://github.com/conan-io/conan-center-index/issues/621

closes https://github.com/conan-io/conan-center-index/issues/6861

---

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
Well, I guess that it won't build in CI without https://github.com/conan-io/conan-center-index/pull/1510. Moreover pyyaml module will be required for build scripts.

**Comment by SpaceIm** [comment]
Here is a summary of the current status for this PR:
- the set of default options is not too far from the pre build "CPU" binaries available on https://pytorch.org/
- with these default options, it builds with many configurations on Windows, Linux & Macos (see https://github.com/SpaceIm/conan-libtorch, even if it fails sometimes due to out of disk space in the container).
- it requires python3 and few external modules (at least pyyaml) at build time to generate several source files. Since cpython recipe is not yet available, libtorch recipe highly relies on python 3.8 (to avoid more external modules) installed on the system.
- I've tested locally with Visual Studio an most options enabled, it works with few limitations:
  - In 1.8.0, `with_vulkan` failed with Visual Studio at least, and after several bug fixes (usage of C++20 features while pytorch is C++14, missing friend classes) I gave up. I can test again in 1.8.1
  - `with_cuda` fails with VC++ 14.28+ (known issue https://github.com/pytorch/pytorch/issues/54382, seems to be a regression in Visual Studio), so I was not able to fully test CUDA build (fails around 1300 on 1800 compilation units).
- there is still something wrong in headers installation due to cmake wrapper (a source_subfolder directory is installed)
- I allow build with CUDA, but CUDA libraries are not propagate in `package_info()`
- if static, several torch libs must be linked with whole archive. I've implemented a tedious workaround in `package_info()` for that.
- On Linux, clang with libc++ fails in configure due to a simple compilation test, I don't know why.
- On Linux, when `with_numa=True` (default value), libnuma must be shared, because numa static can't be linked into shared libs.

**Comment by SpaceIm** [comment]
@jgsogo Would it be possible to install python 3.8 or 3.9 in agents? It would be nice to have libtorch recipe in CCI, even if build_requirements are not perfect.

**Comment by jgsogo** [comment]
> @jgsogo Would it be possible to install python 3.8 or 3.9 in agents? It would be nice to have libtorch recipe in CCI, even if build_requirements are not perfect.

💯 !! We need to upgrade it in all instances (Win/Lin/Mac). It will help us to learn about how some machines are provisioned.

File: recipes/libtorch/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,679 @@
...
+            # should be part of OpenLDAP or packaged separately?
+            raise ConanInvalidConfiguration("lmdb recipe not yet available in CCI")
+        if self.options.get_safe("with_nnpack"):
+            raise ConanInvalidConfiguration("nnpack recipe not yet available in CCI")
+        if self.options.get_safe("with_qnnpack"):
+            self.requires("fp16/cci.20200514")
+            self.requires("fxdiv/cci.20200417")
+            self.requires("psimd/cci.20200517")
+        if self.options.with_xnnpack:
+            self.requires("xnnpack/cci.20210310")
```
Comment (SSE4) [inline recipes/libtorch/all/conanfile.py:232]:
```suggestion
            self.requires("xnnpack/cci.20211210")
```

File: recipes/libtorch/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+            self.requires("glog/0.4.0")
+        if self.options.with_leveldb:
+            self.requires("leveldb/1.22")
+        if self.options.with_lmdb:
+            # should be part of OpenLDAP or packaged separately?
+            raise ConanInvalidConfiguration("lmdb recipe not yet available in CCI")
+        if self.options.get_safe("with_nnpack"):
+            raise ConanInvalidConfiguration("nnpack recipe not yet available in CCI")
+        if self.options.get_safe("with_qnnpack"):
+            self.requires("fp16/cci.20200514")
```
Comment (SSE4) [inline recipes/libtorch/all/conanfile.py:228]:
```suggestion
            self.requires("fp16/cci.20210320")
```

File: recipes/libtorch/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,679 @@
...
+        if self.options.get_safe("with_numa"):
+            self.options["libnuma"].shared = True
+
+    def requirements(self):
+        self.requires("cpuinfo/cci.20201217")
+        self.requires("eigen/3.3.9")
+        self.requires("fmt/7.1.3")
+        self.requires("foxi/cci.20210217")
+        self.requires("onnx/1.8.1")
+        self.requires("protobuf/3.15.5")
```
Comment (SSE4) [inline recipes/libtorch/all/conanfile.py:192]:
```suggestion
        self.requires("protobuf/3.17.1")
```

File: recipes/libtorch/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,679 @@
...
+            self.output.warn("Distributed libtorch is not tested on {} and likely won't work".format(str(self.settings.os)))
+
+        # numa static can't be linked into shared libs.
+        # Because Caffe2_detectron_ops* libs are always shared, we have to force
+        # libnuma shared even if libtorch:shared=False
+        if self.options.get_safe("with_numa"):
+            self.options["libnuma"].shared = True
+
+    def requirements(self):
+        self.requires("cpuinfo/cci.20201217")
```
Comment (SSE4) [inline recipes/libtorch/all/conanfile.py:187]:
```suggestion
        self.requires('cpython/3.8.12')
        self.requires("cpuinfo/cci.20201217")
```

**Comment by SpaceIm** [comment]
@SSE4, adding python in build requirements is not sufficient since few py files invoked at build time depend on few external modules. I'll work again on this PR when I have time.

**Comment by SSE4** [comment]
> @SSE4, adding python in build requirements is not sufficient since few py files invoked at build time depend on few external modules. I'll work again on this PR when I have time.

okay, I just wanted to do some quick tests if `cpython` resolves our infrastructure issues or not.

**Comment by Tumb1eweed** [comment]
> Specify library name and version: **lib/1.0**
> 
> This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!
> 
> #621
> 
> closes #6861
> 
> * [ ]  I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
> * [ ]  I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
> * [ ]  I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
> * [ ]  I've tried at least one configuration locally with the
>   [conan-center hook](https://github.com/conan-io/hooks.git) activated.

A small problem, if build libtorch with "USE_CUDA" option, then it would require nccl as collective communicate library, but nccl
only provide makefile, so when you invoke CMake(self) method, it wouldn't build nccl, end up causing failure. Nccl provide source, so we can forge a nccl recipe from the start.

**Comment by SpaceIm** [comment]
Yes, when I have time. It's a very complex recipe, and it should be ported to conan v2, use cpython recipe as a tool_require (not easy since build step depends on at least one external python dependency), and be updated to latest libtorch version.

**Pull request closed without merging**